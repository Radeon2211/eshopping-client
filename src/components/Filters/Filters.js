import React from 'react';
import PropTypes from 'prop-types';
import { useHistory } from 'react-router-dom';
import queryString from 'query-string';
import Heading from '../UI/Heading/Heading';
import Panel from '../UI/Panel/Panel';

const Filters = (props) => {
  const { products, isListLoading } = props;
  const {
    location: { search },
  } = useHistory();

  let filters = 'Filters';
  if (products) {
    if (products.length <= 0) {
      const queryStringKeys = Object.keys(queryString.parse(search));
      if (queryStringKeys.length > 0) {
        filters = 'Filters';
      } else {
        filters = <Heading variant="h4">Filters are unavailable</Heading>;
      }
      if (queryStringKeys.length === 1 && queryStringKeys.includes('name')) {
        filters = <Heading variant="h4">Filters are unavailable</Heading>;
      }
    }
  }

  return <Panel show={!isListLoading}>{filters}</Panel>;
};

Filters.defaultProps = {
  products: null,
};

Filters.propTypes = {
  products: PropTypes.arrayOf(PropTypes.object),
  isListLoading: PropTypes.bool.isRequired,
};

export default Filters;
