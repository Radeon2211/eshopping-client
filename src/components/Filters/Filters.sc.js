import styled from 'styled-components';

export const Wrapper = styled.div`
  & > *:not(:last-child) {
    margin-bottom: ${({ theme }) => theme.spacings.level3};
  }

  & .select {
    font-size: ${({ theme }) => theme.fontSizes.level2};
    z-index: ${({ theme }) => theme.zIndexes.level2};
  }
`;

export const Toggler = styled.div`
  align-self: flex-end;
  border: 1px solid ${({ theme }) => theme.colors.light4};
  border-radius: 5px;
  cursor: pointer;
  display: flex;
  font-size: ${({ theme }) => theme.fontSizes.level2};
  margin-bottom: ${({ theme }) => theme.spacings.level3};
  padding: ${({ theme }) => theme.spacings.level1} ${({ theme }) => theme.spacings.level2};

  & .label {
    margin: 0 ${({ theme }) => theme.spacings.level2} 0 ${({ theme }) => theme.spacings.level1};
  }
`;

export const Checkboxes = styled.div`
  & > *:not(:last-child) {
    margin-bottom: ${({ theme }) => theme.spacings.level2};
  }

  @media only screen and (max-width: 56.25em) {
    & > *:not(:last-child) {
      margin-bottom: calc(1.5 * ${({ theme }) => theme.spacings.level2});
    }
  }
`;

export const CheckboxBox = styled.div`
  align-items: center;
  display: flex;

  & > label {
    cursor: pointer;
    display: inline-block;
    font-size: ${({ theme }) => theme.fontSizes.level2};
    line-height: ${({ theme }) => theme.lineHeights.level6};
    padding: 0 0 0
      calc(${({ theme }) => theme.spacings.level3} + ${({ theme }) => theme.spacings.level1});
    position: relative;

    &::before {
      align-items: center;
      border-radius: 5px;
      border: 2px solid ${({ theme }) => theme.colors.blue};
      content: '';
      color: ${({ theme }) => theme.colors.blue};
      display: flex;
      height: ${({ theme }) => theme.spacings.level3};
      font-size: ${({ theme }) => theme.fontSizes.level4};
      font-weight: 700;
      justify-content: center;
      left: 0;
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      width: ${({ theme }) => theme.spacings.level3};
    }
  }

  & > input {
    height: 0
    opacity: 0;
    margin: 0;
    width: 0;

    &:checked + label::before {
      content: 'âœ—';
    }
  }
`;
