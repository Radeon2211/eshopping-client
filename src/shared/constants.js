import * as Yup from 'yup';
import { getCountries } from 'country-fns';

export const anyStringRule = Yup.string().min(1).required();

export const userRules = {
  email: Yup.string().email().max(320).trim().required(),
  hideEmail: Yup.bool(),
  username: Yup.string().min(3).max(20).trim().required(),
  password: Yup.string().min(7).max(64).trim().required(),
  firstName: Yup.string().max(60).trim().required(),
  lastName: Yup.string().max(80).trim().required(),
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string()
    .min(5)
    .max(15)
    .matches(/^[\d-]{5,15}$/)
    .trim()
    .required(),
  hidePhone: Yup.bool(),
  street: Yup.string().max(60).trim().required(),
  zipCode: Yup.string().max(12).trim().required(),
  city: Yup.string().max(100).trim().required(),
  country: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
};

export const productConditions = {
  NEW: 'new',
  USED: 'used',
  NOT_APPLICABLE: 'not_applicable',
};

export const productRules = {
  name: Yup.string().max(150).trim().required(),
  price: Yup.number().min(0.01).max(1000000).required(),
  quantity: Yup.number().min(1).max(100000).required(),
  condition: Yup.string()
    .oneOf([productConditions.NEW, productConditions.USED, productConditions.NOT_APPLICABLE])
    .required(),
  description: Yup.string().max(600).trim(),
};

export const listOfAreaCodes = getCountries().map(({ name, dial }) => {
  const finalValue = `+${dial} ${name.split('(')[0].trim()}`;
  return {
    value: dial,
    label: finalValue,
  };
});

export const listOfCountries = getCountries().map(({ name }) => {
  const finalValue = name.split('(')[0].trim();
  return {
    value: finalValue,
    label: finalValue,
  };
});

export const productPages = {
  ALL_PRODUCTS: 'ALL_PRODUCTS',
  MY_PRODUCTS: 'MY_PRODUCTS',
  USER_PRODUCTS: 'USER_PRODUCTS',
};

export const itemTypes = {
  CART: 'CART',
  ORDER: 'ORDER',
  TRANSACTION: 'TRANSACTION',
};

export const listItemTypes = {
  PRODUCT: 'PRODUCT',
  ORDER: 'ORDER',
};

export const orderTypes = {
  PLACED_ORDERS: 'PLACED_ORDERS',
  SELL_HISTORY: 'SELL_HISTORY',
};

export const inputKinds = {
  INPUT: 'INPUT',
  TEXTAREA: 'TEXTAREA',
  SELECT: 'SELECT',
  RADIO: 'RADIO',
};

export const defaultProductsPerPage = 10;

export const defaultOrdersPerPage = 6;

export const defaultAppPath = '/products?p=1';

export const filtersActions = {
  INIT_STATE: 'INIT_STATE',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_MIN_PRICE: 'SET_MIN_PRICE',
  SET_MAX_PRICE: 'SET_MAX_PRICE',
  SET_CONDITION: 'SET_CONDITION',
};

export const sliderPositionsActions = {
  SET_BOTH: 'SET_BOTH',
  SET_LEFT: 'SET_LEFT',
  SET_RIGHT: 'SET_RIGHT',
};

export const modalTypes = {
  ABOUT_WEBSITE: 'ABOUT_WEBSITE',
  ADD_ADMIN: 'ADD_ADMIN',
  ADD_PRODUCT: 'ADD_PRODUCT',
  BUY_PRODUCTS: 'BUY_PRODUCTS',
  CART_ITEM_ADDED: 'CART_ITEM_ADDED',
  CHANGE_NAME: 'CHANGE_NAME',
  CHANGE_EMAIL: 'CHANGE_EMAIL',
  CHANGE_USERNAME: 'CHANGE_USERNAME',
  CHANGE_PHONE_NUMBER: 'CHANGE_PHONE_NUMBER',
  CHANGE_ADDRESS: 'CHANGE_ADDRESS',
  CHANGE_PASSWORD: 'CHANGE_PASSWORD',
  CHANGE_CONTACTS: 'CHANGE_CONTACTS',
  CHANGE_DELIVERY_ADDRESS: 'CHANGE_DELIVERY_ADDRESS',
  CLEAR_CART: 'CLEAR_CART',
  DELETE_ACCOUNT: 'DELETE_ACCOUNT',
  DELETE_PRODUCT: 'DELETE_PRODUCT',
  EDIT_PRODUCT: 'EDIT_PRODUCT',
  LOGIN: 'LOGIN',
  REMOVE_ADMIN: 'REMOVE_ADMIN',
  RESET_PASSWORD: 'RESET_PASSWORD',
  PENDING_USER_INFO: 'PENDING_USER_INFO',
  SEND_VERIFICATION_LINK: 'SEND_VERIFICATION_LINK',
  SIGNUP: 'SIGNUP',
};

export const sortProductsOptions = [
  { value: undefined, label: 'Default sorting' },
  { value: 'price:asc', label: `Price - ascending` },
  { value: 'price:desc', label: 'Price - descending' },
  { value: 'name:asc', label: 'Name - A to Z' },
  { value: 'name:desc', label: 'Name - Z to A' },
];

export const sortOrdersOptions = [
  { value: 'createdAt:desc', label: 'Date - from newest' },
  { value: 'createdAt:asc', label: 'Date - from oldest' },
  { value: 'overallPrice:asc', label: `Price - ascending` },
  { value: 'overallPrice:desc', label: 'Price - descending' },
];

export const singleInfoNames = {
  USERNAME: 'Username',
  NAME: 'Name',
  EMAIL: 'Email',
  PHONE_NUMBER: 'Phone number',
  ADDRESS: 'Address',
  CONTACTS: 'Contacts',
};

export const updateCartActions = {
  INCREMENT: 'INCREMENT',
  DECREMENT: 'DECREMENT',
  NUMBER: 'NUMBER',
};

export const defaultErrorMessage = 'Something went wrong';

export const productsPerPageControllerOptions = [
  { value: 1, label: '1' },
  { value: 2, label: '2' },
  { value: 3, label: '3' },
  { value: 4, label: '4' },
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 25, label: '25' },
];

export const defaultScrollToConfig = {
  top: 0,
  left: 0,
  behavior: 'smooth',
};

export const userStatuses = {
  ACTIVE: 'active',
  PENDING: 'pending',
};
