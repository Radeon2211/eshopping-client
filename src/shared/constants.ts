import * as Yup from 'yup';
import countryCodes from 'country-codes-list';
import { ProductCondition, ProductPhotoFieldValue } from './types/enums';

export const anyStringRule = Yup.string().min(1).required();

export const userRules = {
  email: Yup.string().email().max(320).trim().required(),
  hideEmail: Yup.bool(),
  username: Yup.string().min(3).max(20).trim().required(),
  password: Yup.string().min(7).max(64).trim().required(),
  firstName: Yup.string().max(60).trim().required(),
  lastName: Yup.string().max(80).trim().required(),
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string()
    .min(5)
    .max(15)
    .matches(/^[\d-]{5,15}$/)
    .trim()
    .required(),
  hidePhone: Yup.bool(),
  street: Yup.string().max(60).trim().required(),
  zipCode: Yup.string().max(12).trim().required(),
  city: Yup.string().max(100).trim().required(),
  country: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
};

export const productRules = {
  name: Yup.string().max(150).trim().required(),
  price: Yup.number().min(0.01).max(1000000).required(),
  quantity: Yup.number().min(1).max(100000).required(),
  condition: Yup.string()
    .oneOf([ProductCondition.NEW, ProductCondition.USED, ProductCondition.NOT_APPLICABLE])
    .required(),
  description: Yup.string().max(600).trim(),
  photo: Yup.mixed()
    .test('is-error', (value) => value !== ProductPhotoFieldValue.ERROR)
    .nullable(),
};

export const listOfAreaCodes = countryCodes.customArray({
  value: '{countryCallingCode}',
  label: '+{countryCallingCode} {countryNameEn}',
});

export const listOfCountries = countryCodes.customArray({
  value: '{countryNameEn}',
  label: '{countryNameEn}',
});

export const itemTypes = {
  CART: 'CART',
  ORDER: 'ORDER',
  TRANSACTION: 'TRANSACTION',
};

export const listItemTypes = {
  PRODUCT: 'PRODUCT',
  ORDER: 'ORDER',
};

export const inputKinds = {
  INPUT: 'INPUT',
  TEXTAREA: 'TEXTAREA',
  SELECT: 'SELECT',
  RADIO: 'RADIO',
};

export const defaultProductsPerPage = 10;

export const defaultOrdersPerPage = 6;

export const defaultAppPath = '/products?p=1';

export const filtersActions = {
  INIT_STATE: 'INIT_STATE',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_MIN_PRICE: 'SET_MIN_PRICE',
  SET_MAX_PRICE: 'SET_MAX_PRICE',
  SET_CONDITION: 'SET_CONDITION',
};

export const sliderPositionsActions = {
  SET_BOTH: 'SET_BOTH',
  SET_LEFT: 'SET_LEFT',
  SET_RIGHT: 'SET_RIGHT',
};

export const sortProductsOptions = [
  { value: undefined, label: 'Default sorting' },
  { value: 'price:asc', label: 'Price - ascending' },
  { value: 'price:desc', label: 'Price - descending' },
  { value: 'name:asc', label: 'Name - A to Z' },
  { value: 'name:desc', label: 'Name - Z to A' },
];

export const sortOrdersOptions = [
  { value: 'createdAt:desc', label: 'Date - from newest' },
  { value: 'createdAt:asc', label: 'Date - from oldest' },
  { value: 'overallPrice:asc', label: 'Price - ascending' },
  { value: 'overallPrice:desc', label: 'Price - descending' },
];

export const singleInfoNames = {
  USERNAME: 'Username',
  NAME: 'Name',
  EMAIL: 'Email',
  PHONE_NUMBER: 'Phone number',
  ADDRESS: 'Address',
  CONTACTS: 'Contacts',
};

export const defaultErrorMessage = 'Something went wrong';

export const productsPerPageControllerOptions = [
  { value: 1, label: '1' },
  { value: 2, label: '2' },
  { value: 3, label: '3' },
  { value: 4, label: '4' },
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 25, label: '25' },
];

export const defaultScrollToConfig: ScrollToOptions = {
  top: 0,
  left: 0,
  behavior: 'smooth',
};
