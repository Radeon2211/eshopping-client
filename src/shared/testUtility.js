// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const historyPageNum = (pageNumber) => ({
  listen: jest.fn(),
  location: { search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push: jest.fn(),
});

export const propsPagination = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  maxQuantityPerPage: 2,
});
