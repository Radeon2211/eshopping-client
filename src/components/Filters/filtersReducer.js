import { updateObject } from '../../shared/utility';
import { filtersActions, sortProductsOptions } from '../../shared/constants';

export const filtersInitialState = {
  sortBy: sortProductsOptions[0],
  minPrice: undefined,
  maxPrice: undefined,
  condition: {
    new: false,
    used: false,
    not_applicable: false,
  },
};

export const filtersReducer = (state = filtersInitialState, action) => {
  switch (action.type) {
    case filtersActions.INIT_STATE:
      return updateObject(state, action.payload);
    case filtersActions.SET_SORT_BY:
      return updateObject(state, { sortBy: action.sortBy });
    case filtersActions.SET_MIN_PRICE:
      return updateObject(state, { minPrice: action.minPrice });
    case filtersActions.SET_MAX_PRICE:
      return updateObject(state, { maxPrice: action.maxPrice });
    case filtersActions.SET_CONDITION:
      return updateObject(state, { condition: { ...state.condition, ...action.condition } });
    default:
      return state;
  }
};
