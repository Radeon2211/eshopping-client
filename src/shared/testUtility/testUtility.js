// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';
import { v4 as uuidv4 } from 'uuid';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../store/reducers/rootReducer';
import { productConditions } from '../constants';

export const defaultDeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: 'active',
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const createHistoryPageNumber = (pageNumber = 1, push = jest.fn()) => ({
  listen: jest.fn(),
  location: { pathname: '/products', search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push,
});

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
  quantitySold = 0,
  buyerQuantity = 0,
  description = '',
  condition = productConditions.NEW,
) => ({
  _id: productId || uuidv4(),
  name,
  price,
  quantity,
  quantitySold,
  buyerQuantity,
  description,
  condition,
  photo,
  seller: {
    username: sellerUsername,
  },
  __v: 0,
  createdAt: '2021-02-10T19:10:38.872Z',
  updatedAt: '2021-02-10T19:10:38.872Z',
});

export const createCartItem = (
  sellerUsername = 'username',
  quantity = 1,
  productId,
  price = 2,
  name = 'product name',
  productQuantity = 2,
  photo = false,
) => {
  const product = createProductItem(
    productId || uuidv4(),
    sellerUsername,
    productQuantity,
    price,
    name,
    photo,
  );

  return {
    _id: uuidv4(),
    quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
) => ({
  _id: productId || uuidv4(),
  name,
  photo,
  price,
  quantity,
  seller: {
    username: sellerUsername,
  },
});

export const createOrder = (
  products = [createTransactionAndOrderProdItem()],
  orderId = uuidv4(),
  sellerUsername,
  buyerUsername,
  overallPrice,
  createdAt,
  sellerEmail,
  sellerPhone,
) => {
  const seller = sellerUsername
    ? {
        username: sellerUsername,
        email: sellerEmail,
        phone: sellerPhone,
      }
    : null;

  const buyer = buyerUsername
    ? {
        username: buyerUsername,
      }
    : null;

  return {
    _id: orderId,
    seller,
    buyer,
    products,
    overallPrice,
    deliveryAddress: defaultDeliveryAddress,
    createdAt,
    updatedAt: createdAt,
    __v: 0,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...defaultState.product,
      ...initialProduct,
    },
    ui: {
      ...defaultState.ui,
      ...initialUI,
    },
  };
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

export const checkReqMethodAndURL = (moxiosInstance, method, url, reqNumber = 0) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests[reqNumber - 1];
    return method === prevRequest.config.method && url === prevRequest.config.url;
  }
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState,
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};
