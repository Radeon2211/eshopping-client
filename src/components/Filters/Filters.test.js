import React from 'react';
import { mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import thunk from 'redux-thunk';
import Filters from './Filters';
import * as SC from './Filters.sc';
import { checkProps } from '../../shared/testUtility';
import theme from '../../styled/theme';
import { sortOptions, filtersActions } from '../../shared/constants';
import { filtersReducer, filtersInitialState } from './filtersReducer';

const mockStore = configureMockStore([thunk]);

const defaultStore = mockStore({
  product: {
    minPrice: 10,
    maxPrice: 100,
  },
});

const createProps = (products, isDataLoading = false) => ({
  products,
  isDataLoading,
});

const defaultProps = createProps([{ id: 1 }]);

const simulateTogglerClick = (wrapper) => {
  const toggler = wrapper.find(SC.Toggler);
  toggler.simulate('click');
};

const setUp = (props, search = '', push = jest.fn()) => {
  const history = {
    listen: jest.fn(),
    createHref: jest.fn(),
    location: { pathname: '/products', search },
    push,
  };
  return mount(
    <Router history={history}>
      <Provider store={defaultStore}>
        <ThemeProvider theme={theme}>
          <Filters {...props} />
        </ThemeProvider>
      </Provider>
    </Router>,
  );
};

describe('<Filters />', () => {
  describe('Check prop types', () => {
    it('Should NOT throw a warning', () => {
      expect(checkProps(Filters, defaultProps)).toBeUndefined();
    });

    it('Should throw a warning', () => {
      expect(checkProps(Filters, {})).not.toBe(null);
    });
  });

  describe(`Check if renders correctly`, () => {
    it('Should render <SC.Wrapper /> and submit button should NOT be disabled', () => {
      const wrapper = setUp(defaultProps);
      simulateTogglerClick(wrapper);
      expect(wrapper.find(SC.Wrapper)).toHaveLength(1);
      expect(wrapper.find('[data-test="filters-submit-btn"]').first().prop('disabled')).toBe(false);
    });

    it('Should submit button be disabled', () => {
      const props = createProps([{ id: 1 }], true);
      const wrapper = setUp(props);
      simulateTogglerClick(wrapper);
      expect(wrapper.find('[data-test="filters-submit-btn"]').first().prop('disabled')).toBe(true);
    });

    it('Should NOT render <SC.Wrapper /> and unavailable heading', () => {
      const wrapper = setUp(defaultProps);
      expect(wrapper.find(SC.Wrapper)).toHaveLength(0);
      expect(wrapper.find('[data-test="unavailable-heading"]')).toHaveLength(0);
    });

    it('Should render unavailable heading', () => {
      const props = createProps([]);
      const wrapper = setUp(props);
      simulateTogglerClick(wrapper);
      expect(wrapper.find('[data-test="unavailable-heading"]').length).toBeGreaterThan(0);
    });

    it('Should NOT render unavailable heading if in url is p and sth else than name', () => {
      const props = createProps([]);
      const wrapper = setUp(props, '?p=1&minPrice=10');
      simulateTogglerClick(wrapper);
      expect(wrapper.find('[data-test="unavailable-heading"]')).toHaveLength(0);
    });

    it('Should render unavailable heading if in url is name', () => {
      const props = createProps([]);
      const wrapper = setUp(props, '?name=test-name');
      simulateTogglerClick(wrapper);
      expect(wrapper.find('[data-test="unavailable-heading"]').length).toBeGreaterThan(0);
    });

    it('Should NOT render unavailable heading if in url is sth expect p and name', () => {
      const props = createProps([]);
      const wrapper = setUp(props, '?minPrice=10');
      simulateTogglerClick(wrapper);
      expect(wrapper.find('[data-test="unavailable-heading"]')).toHaveLength(0);
    });
  });

  describe(`Check if controls' values are correct`, () => {
    it('Should condition checkboxes be checked', () => {
      const wrapper = setUp(defaultProps, '?p=1&condition=new,used,not_applicable');
      simulateTogglerClick(wrapper);
      expect(wrapper.find('#new').prop('checked')).toBe(true);
      expect(wrapper.find('#used').prop('checked')).toBe(true);
      expect(wrapper.find('#not_applicable').prop('checked')).toBe(true);
    });

    it('Should NOT condition checkboxes be checked', () => {
      const wrapper = setUp(defaultProps);
      simulateTogglerClick(wrapper);
      expect(wrapper.find('#new').prop('checked')).toBe(false);
      expect(wrapper.find('#used').prop('checked')).toBe(false);
      expect(wrapper.find('#not_applicable').prop('checked')).toBe(false);
    });

    it('Should sort by price ascending', () => {
      const wrapper = setUp(defaultProps, '?p=1&sortBy=price:asc');
      simulateTogglerClick(wrapper);
      expect(wrapper.find('.select').first().prop('value')).toBe(sortOptions[1]);
    });

    it('Should sort by price descending', () => {
      const wrapper = setUp(defaultProps, '?p=1&sortBy=price:desc');
      simulateTogglerClick(wrapper);
      expect(wrapper.find('.select').first().prop('value')).toBe(sortOptions[2]);
    });

    it('Should sort by name ascending', () => {
      const wrapper = setUp(defaultProps, '?p=1&sortBy=name:asc');
      simulateTogglerClick(wrapper);
      expect(wrapper.find('.select').first().prop('value')).toBe(sortOptions[3]);
    });

    it('Should sort by name descending', () => {
      const wrapper = setUp(defaultProps, '?p=1&sortBy=name:desc');
      simulateTogglerClick(wrapper);
      expect(wrapper.find('.select').first().prop('value')).toBe(sortOptions[4]);
    });
  });

  describe('Check filtersReducer', () => {
    it('Should return default state', () => {
      expect(filtersReducer(undefined, {})).toEqual(filtersInitialState);
    });

    it('Should return new state after INIT_STATE', () => {
      const payload = {
        sortBy: sortOptions[1],
        minPrice: 10,
        maxPrice: 100,
        condition: {
          new: true,
          used: true,
          not_applicable: false,
        },
      };
      expect(
        filtersReducer(undefined, {
          type: filtersActions.INIT_STATE,
          payload,
        }),
      ).toEqual(payload);
    });

    it('Should return new state after SET_SORT_BY', () => {
      expect(
        filtersReducer(undefined, {
          type: filtersActions.SET_SORT_BY,
          sortBy: sortOptions[2],
        }),
      ).toEqual({
        ...filtersInitialState,
        sortBy: sortOptions[2],
      });
    });

    it('Should return new state after SET_MIN_PRICE', () => {
      expect(
        filtersReducer(undefined, {
          type: filtersActions.SET_MIN_PRICE,
          minPrice: 10,
        }),
      ).toEqual({
        ...filtersInitialState,
        minPrice: 10,
      });
    });

    it('Should return new state after SET_MAX_PRICE', () => {
      expect(
        filtersReducer(undefined, {
          type: filtersActions.SET_MAX_PRICE,
          maxPrice: 100,
        }),
      ).toEqual({
        ...filtersInitialState,
        maxPrice: 100,
      });
    });

    it('Should return new state after SET_CONDITION', () => {
      expect(
        filtersReducer(undefined, {
          type: filtersActions.SET_CONDITION,
          condition: {
            new: true,
          },
        }),
      ).toEqual({
        ...filtersInitialState,
        condition: {
          ...filtersInitialState.condition,
          new: true,
        },
      });
    });
  });
});
