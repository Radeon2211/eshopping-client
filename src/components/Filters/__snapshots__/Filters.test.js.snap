// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filters /> check how renders should render everything correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsTCUz drXtAl"
    data-testid="Filters-toggler"
  >
    <div
      class="sc-dQppl hTXijc"
      color=""
    >
      <svg>
        filters.svg
      </svg>
    </div>
    <span
      class="label"
    >
      Filters
    </span>
    <div
      class="sc-dQppl juhMHn"
      color=""
    >
      <svg>
        arrow.svg
      </svg>
    </div>
  </div>
  <section
    class="sc-eCssSg cuFRxW"
    data-testid="Filters"
  >
    <div
      class="sc-bdfBwQ XTQGb"
    >
      <div
        class="select css-2b097c-container"
      >
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-yk16xz-control"
        >
          <div
            class=" css-g1d714-ValueContainer"
          >
            <div
              class=" css-1uccc91-singleValue"
            >
              Default sorting
            </div>
            <input
              aria-autocomplete="list"
              class="css-wmatm6-dummyInput-DummyInput"
              id="react-select-2-input"
              readonly=""
              tabindex="0"
              value=""
            />
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="sortBy"
          type="hidden"
          value=""
        />
      </div>
      <div
        class="sc-jrAGrp fPUvfA"
      >
        <div
          class="sc-kEjbxe kHvlHc"
        >
          <span
            class="sc-jSgupP hYzbrm"
            color=""
            display="inline-block"
            overflow="visible"
            size="3"
            spacing="0px"
            transform=""
          >
            Price ($)
          </span>
          <div
            class="inputs"
          >
            <input
              class="sc-crrsfI dYUtiP"
              data-testid="NumberInput-minPrice"
              name="minPrice"
              type="number"
              value="10"
            />
            <span
              class="inputs-gap"
            >
              â€”
            </span>
            <input
              class="sc-crrsfI dYUtiP"
              data-testid="NumberInput-maxPrice"
              name="maxPrice"
              type="number"
              value="100"
            />
          </div>
        </div>
        <div
          class="sc-iqHYGH cknxsw"
        >
          <input
            class="input min"
            data-testid="PriceSlider-price-range-min"
            max="100"
            min="10"
            name="minPrice"
            step="0.1"
            type="range"
            value="10"
          />
          <input
            class="input max"
            data-testid="PriceSlider-price-range-max"
            max="100"
            min="10"
            name="maxPrice"
            step="0.1"
            type="range"
            value="100"
          />
          <div
            class="track"
          />
          <div
            class="range"
          />
          <div
            class="thumb left"
          />
          <div
            class="thumb right"
          />
        </div>
      </div>
      <div
        class="sc-dlfnbm jDofXQ"
      >
        <span
          class="sc-jSgupP hYzbrm"
          color=""
          display="inline-block"
          overflow="visible"
          size="3"
          spacing="0px"
          transform=""
        >
          Condition
        </span>
        <div
          class="sc-hKgILt jUSMYX"
        >
          <input
            data-testid="Filters-checkbox-new"
            id="new"
            name="new"
            type="checkbox"
          />
          <label
            for="new"
          >
            new
          </label>
        </div>
        <div
          class="sc-hKgILt jUSMYX"
        >
          <input
            data-testid="Filters-checkbox-used"
            id="used"
            name="used"
            type="checkbox"
          />
          <label
            for="used"
          >
            used
          </label>
        </div>
        <div
          class="sc-hKgILt jUSMYX"
        >
          <input
            data-testid="Filters-checkbox-not-applicable"
            id="not_applicable"
            name="not_applicable"
            type="checkbox"
          />
          <label
            for="not_applicable"
          >
            not applicable
          </label>
        </div>
      </div>
      <button
        class="sc-gKsewC gFbfGD"
        color="blue"
        data-testid="Filters-submit-btn"
        type="button"
      >
        Filter
      </button>
    </div>
  </section>
</DocumentFragment>
`;
